ls1intum/Artemis
# `Programming exercises:` Allow to publish example solutions for programming exercises on a specific date

### Checklist
#### General


#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context


Continuation PR of https://github.com/ls1intum/Artemis/pull/4328 to resolve https://github.com/ls1intum/Artemis/issues/16. Allows students to download example solution repositories after a specified date.
### Description

This allows the students to download the solution repository of a given programming exercise as a zip file when the `Example Solution Publication Date` is set and before now. `Example Solution Publication Date` is present in the `BaseExercise` class (and in the database) since https://github.com/ls1intum/Artemis/pull/4328 but it was not present for the Programming Exercise detail and editing page at the client so it is added to the Programming Exercise Lifecycle Component Different from the repositories downloadable by the instructors at the course detail page, student zip files will have their `.git` directories removed in order to not leak unintended information to the students.
### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise

Instructor:

1. Log in to Artemis
2. Navigate to Course Administration
3. Go to the Edit page of the prerequisite Programming Exercise.
4. Ensure current `Example Solution Publication Date` is empty or after current time.

Student:

5. Log in to Artemis
6. Navigate to the prerequisite Programming Exercise's detail page.
7. Ensure example solution download button is not visible.

Instructor:

8. Change `Example Solution Publication Date` of the exercise to a time before current time.

Student:

9. Ensure example solution download button with text "Download Repo" is visible (See screenshots).
10. Download the file and check the downloaded zip file contains the files from the solution repository but does not contain `.git` directory.

Instructor:

12. Import an existing Programming Exercise with `Example Solution Publication Date` set to a non-null date.
13. Ensure new exercise has `Example Solution Publication Date` set to null.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage





| Class/File | Branch | Line |
|------------|-------:|-----:|
|ZipFileService.java| 64% | 94%|
|GitService.java| 76% | 86%|
|ProgrammingExerciseExportService.java| 63% | 90%|
|ProgrammingExerciseExportImportResource.java| 80% | 98%|
|exercise.service.ts| 66.66% | 80.95%|
|course-exercise-details.component.ts| 59.51% | 79.92%|

### Screenshots



![image](https://user-images.githubusercontent.com/18427209/160654897-9ed7d61f-0e55-4c74-a4d0-6f2b396c6502.png)

![image](https://user-images.githubusercontent.com/18427209/160654575-aaf84be1-03ce-4e43-9fe5-131107f8c7a8.png)

![image](https://user-images.githubusercontent.com/18427209/160654651-60f5c805-f986-4e54-89e5-ebe07122ca12.png)
